core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -805,13 +814,12 @@ private[spark] class BlockManager(
           }
         }
       } finally {
-        // If we failed in putting the block to memory/disk, notify other possible readers
-        // that it has failed, and then remove it from the block info map.
-        if (!marked) {
-          // Note that the remove must happen before markFailure otherwise another thread
-          // could've inserted a new BlockInfo before we remove it.
-          blockInfo.remove(blockId)
-          putBlockInfo.markFailure()
+        if (!blockWasSuccessfullyStored) {
+          // Guard against the fact that MemoryStore might have already removed the block if the
+          // put() failed and the block could not be dropped to disk.
+          if (blockInfoManager.lockForWriting(blockId, blocking = false).isDefined) {
```

Comment:
How can you fail to lock for writing here? The lock was acquired on line 722 right?